File: pypi/src/sendsoon_mcp/tools/ip_lookup.py
"""ip_lookup MCP tool."""

from __future__ import annotations

from typing import TYPE_CHECKING, Annotated, Any

from pydantic import Field

from sendsoon_mcp.validation import validate_public_ip

if TYPE_CHECKING:
    from mcp.server.fastmcp import FastMCP

    from sendsoon_mcp.client import SendSoonClient

TOOL_TITLE = "IP Lookup"
TOOL_DESCRIPTION = (
    "Look up geolocation and ISP info for a public IPv4 or IPv6 address via SendSoon"
)


def register(mcp: FastMCP, client: SendSoonClient) -> None:
    @mcp.tool(name="ip_lookup", title=TOOL_TITLE, description=TOOL_DESCRIPTION)
    async def ip_lookup(
        ip: Annotated[str, Field(description="Public IPv4 or IPv6 address to look up")],
    ) -> dict[str, Any]:
        """Look up public IP geolocation and ISP info."""
        trimmed = ip.strip()
        validation_error = validate_public_ip(trimmed)
        if validation_error:
            return {"success": False, "error": validation_error.to_dict()}
        return await client.ip_lookup({"ip": trimmed})

File: npm/src/tools/ip_lookup.ts
import {
  ipLookupFailureResult,
  validatePublicIp,
  type IpLookupResult,
  type SendSoonClient,
} from '../core/index.js';
import * as z from 'zod/v4';
import { formatToolResult } from './format.js';

const ipLookupInputSchema = {
  ip: z.string().trim().min(1).describe('Public IPv4 or IPv6 address to look up'),
};

const ipLookupOutputSchema = {
  success: z.boolean(),
  ip: z.string().optional(),
  ip2region: z
    .object({
      country: z.string(),
      countryCode: z.string(),
      region: z.string(),
      city: z.string(),
      postalCode: z.string(),
      timezone: z.string(),
      latitude: z.number().nullable(),
      longitude: z.number().nullable(),
    })
    .optional(),
  network: z
    .object({
      isp: z.string(),
      asn: z.string(),
      organization: z.string(),
    })
    .optional(),
  source: z.string().optional(),
  error: z
    .object({
      code: z.string(),
      message: z.string(),
      retryable: z.boolean(),
    })
    .optional(),
};

export type IpLookupInput = {
  ip: string;
};

export type IpLookupOutput = IpLookupResult;

export const ipLookupToolDefinition = {
  name: 'ip_lookup' as const,
  config: {
    title: 'IP 归属查询',
    description:
      'Look up geolocation and ISP info for a public IPv4 or IPv6 address via SendSoon',
    inputSchema: ipLookupInputSchema,
    outputSchema: ipLookupOutputSchema,
  },
  createHandler(client: SendSoonClient) {
    return async (input: IpLookupInput) => {
      const ip = input.ip.trim();
      const validationError = validatePublicIp(ip);
      if (validationError) {
        return formatToolResult(ipLookupFailureResult(validationError));
      }

      const result = await client.ipLookup({ ip });
      return formatToolResult(result);
    };
  },
};
